src/rust/engine/fs/src/lib.rs

```diff
@@ -236,6 +237,47 @@ impl PathStat {
     }
 }
 
+#[derive(Clone, Debug, DeepSizeOf, Eq, PartialEq)]
+pub enum PathMetadataKind {
+    File,
+    Directory,
+    Symlink,
+}
+
+/// Expanded version of `Stat` when access to additional filesystem attributes is necessary.
+#[derive(Clone, Debug, DeepSizeOf, Eq, PartialEq)]
+pub struct PathMetadata {
```

Comment:
My pyo3-fu is weak, but isn't there an idiomatic way to generate the Python-side class from this struct? Do we actually need two parallel structs? Or does the engine require "real" Python dataclasses?

Reply:
I agree; I would rather just use a Python side struct for all of this. Even using the `kwargs` constructor is unwiedly (but safer than positional args used by some other constructor logic since reordering fields won't break the kwargs constructor call).

Reply:
I switched to using `PathMetadata` as a pyo3-exported Rust struct. Simplified much of the PR.

Reply:
I switched to using `PathMetadata` as a pyo3-exported Rust struct. Simplified much of the PR.